frescobaldi_app/vcs/__init__.py

```diff
@@ -37,6 +37,8 @@ class VCSError(Exception):
     'git': "vcs.gitrepo"
 }
 
+_app_is_git_controlled = False  # until proven otherwise
```

Comment:
`vcs.app_is_git_controlled()` relies on the fact that the variable is initially undefined. Initializing it with a value means that `vcs.app_is_git_controlled()` always returns that initial value. That makes all the git-related functionality (like the `Git` menu included in the main menu when running Frescobaldi from a git repository rather than from an installed package) unavailable.

Do you remember what the change was needed for?

Reply:
Python 3.12 on Windows raises an exception if it's undefined.

Reply:
That's strange. No issues running without the line on Python 3.12(.4) on Linux. There shouldn't be differences in behavior of general language constructs between OSes.

Would you share error output of the exception?

Reply:
Here's what commenting that line does on mine. I'd only skimmed the traceback before because this was just one of many exceptions blocking the main window from loading early on. Looking more closely, the actual problem may be the Git executable not being where my copy of Python can find it. (I use Git from [MSYS2](https://www.msys2.org/), which is a separate environment from the regular Windows command line.)
```
Traceback (most recent call last):
  File "D:\Users\Benjamin Johnson\Documents\Source Code\frescobaldi\frescobaldi_app\plugin.py", line 79, in instance
    return _instances[cls][obj]
           ~~~~~~~~~~^^^^^
  File "D:\Users\Benjamin Johnson\AppData\Local\Programs\Python\Python312\Lib\weakref.py", line 415, in __getitem__
    return self.data[ref(key)]
           ~~~~~~~~~^^^^^^^^^^
KeyError: <weakref at 0x000001C3FDDBB1F0; to 'type' at 0x000001C3FA2BA5E0 (PanelManager)>

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "D:\Users\Benjamin Johnson\Documents\Source Code\frescobaldi\frescobaldi_app\vcs\__init__.py", line 72, in app_is_git_controlled
    return _app_is_git_controlled
           ^^^^^^^^^^^^^^^^^^^^^^
NameError: name '_app_is_git_controlled' is not defined. Did you mean: 'app_is_git_controlled'?

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "D:\Users\Benjamin Johnson\Documents\Source Code\frescobaldi\frescobaldi", line 30, in <module>
    main()
  File "D:\Users\Benjamin Johnson\Documents\Source Code\frescobaldi\frescobaldi_app\__main__.py", line 231, in main
    win = mainwindow.MainWindow()
          ^^^^^^^^^^^^^^^^^^^^^^^
  File "D:\Users\Benjamin Johnson\Documents\Source Code\frescobaldi\frescobaldi_app\mainwindow.py", line 142, in __init__
    self.createMenus()
  File "D:\Users\Benjamin Johnson\Documents\Source Code\frescobaldi\frescobaldi_app\mainwindow.py", line 1130, in createMenus
    menu.createMenus(self)
  File "D:\Users\Benjamin Johnson\Documents\Source Code\frescobaldi\frescobaldi_app\menu.py", line 61, in createMenus
    m.addMenu(menu_file(mainwindow))
              ^^^^^^^^^^^^^^^^^^^^^
  File "D:\Users\Benjamin Johnson\Documents\Source Code\frescobaldi\frescobaldi_app\menu.py", line 95, in menu_file
    m.addMenu(snippet.menu.TemplateMenu(mainwindow))
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "D:\Users\Benjamin Johnson\Documents\Source Code\frescobaldi\frescobaldi_app\snippet\menu.py", line 149, in __init__
    self.addAction(self.tool().actionCollection.templates_manage)
                   ^^^^^^^^^^^
  File "D:\Users\Benjamin Johnson\Documents\Source Code\frescobaldi\frescobaldi_app\snippet\menu.py", line 58, in tool
    return panelmanager.manager(self.mainwindow()).snippettool
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "D:\Users\Benjamin Johnson\Documents\Source Code\frescobaldi\frescobaldi_app\panelmanager.py", line 38, in manager
    return PanelManager.instance(mainwindow)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "D:\Users\Benjamin Johnson\Documents\Source Code\frescobaldi\frescobaldi_app\plugin.py", line 84, in instance
    result.__init__(obj)
  File "D:\Users\Benjamin Johnson\Documents\Source Code\frescobaldi\frescobaldi_app\panelmanager.py", line 86, in __init__
    if app.is_git_controlled() or QSettings().value("experimental-features", False, bool):
       ^^^^^^^^^^^^^^^^^^^^^^^
  File "D:\Users\Benjamin Johnson\Documents\Source Code\frescobaldi\frescobaldi_app\app.py", line 304, in is_git_controlled
    import vcs
  File "D:\Users\Benjamin Johnson\Documents\Source Code\frescobaldi\frescobaldi_app\vcs\__init__.py", line 96, in <module>
    app_is_git_controlled()
  File "D:\Users\Benjamin Johnson\Documents\Source Code\frescobaldi\frescobaldi_app\vcs\__init__.py", line 76, in app_is_git_controlled
    if is_available('git'):
       ^^^^^^^^^^^^^^^^^^^
  File "D:\Users\Benjamin Johnson\Documents\Source Code\frescobaldi\frescobaldi_app\vcs\__init__.py", line 59, in is_available
    return mod.Repo.vcs_available()
           ^^^^^^^^^^^^^^^^^^^^^^^^
  File "D:\Users\Benjamin Johnson\Documents\Source Code\frescobaldi\frescobaldi_app\vcs\gitrepo.py", line 86, in vcs_available
    cls.run_command('--version')
  File "D:\Users\Benjamin Johnson\Documents\Source Code\frescobaldi\frescobaldi_app\vcs\gitrepo.py", line 68, in run_command
    pr = subprocess.Popen(cmd, cwd=dir,
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "D:\Users\Benjamin Johnson\AppData\Local\Programs\Python\Python312\Lib\subprocess.py", line 1026, in __init__
    self._execute_child(args, executable, preexec_fn, close_fds,
  File "D:\Users\Benjamin Johnson\AppData\Local\Programs\Python\Python312\Lib\subprocess.py", line 1538, in _execute_child
    hp, ht, pid, tid = _winapi.CreateProcess(executable, args,
                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
FileNotFoundError: [WinError 2] The system cannot find the file specified
```

Reply:
Good news is that the issue is not Windows-specific. I was able to replicate the error on Linux by modifying current master (be880e16) like this

```diff
--- a/frescobaldi_app/vcs/gitrepo.py
+++ b/frescobaldi_app/vcs/gitrepo.py
@@ -62,7 +62,7 @@ class Repo(AbstractVCSRepo):
         s = QSettings()
         s.beginGroup("helper_applications")
         git_cmd = s.value("git", "git", str)
-        git_cmd = git_cmd if git_cmd else "git"
+        git_cmd = 'gyt'
         cmd = [git_cmd, cmd]
         cmd.extend(args)
         pr = subprocess.Popen(cmd, cwd=dir,
```

Running Frescobaldi from a cloned git repo while not having git available on the system seems to be a rare scenario, so far not accounted for in the code.

I believe the correct way to address it is

```diff
--- a/frescobaldi_app/vcs/gitrepo.py
+++ b/frescobaldi_app/vcs/gitrepo.py
@@ -65,10 +65,13 @@ class Repo(AbstractVCSRepo):
         git_cmd = git_cmd if git_cmd else "git"
         cmd = [git_cmd, cmd]
         cmd.extend(args)
-        pr = subprocess.Popen(cmd, cwd=dir,
-                              stdout=subprocess.PIPE,
-                              stderr=subprocess.PIPE,
-                              universal_newlines=True)
+        try:
+            pr = subprocess.Popen(cmd, cwd=dir,
+                                  stdout=subprocess.PIPE,
+                                  stderr=subprocess.PIPE,
+                                  universal_newlines=True)
+        except FileNotFoundError as e:
+            raise GitError(str(e))
         (out, error) = pr.communicate()
         if error:
             raise GitError(error)
```

Reply:
Makes sense to me.

Reply:
By the way, thanks again for all you're doing with this @igneus. Your testing, fixes, and critiques have been a tremendous help. I appreciate your care and attention to detail, and hope my occasional "just get it to run" tendencies haven't caused you (and the folks upstream) too much trouble.

Reply:
Quite the opposite. Your upgrade (hopefully) saved Frescobaldi from slow death and it's my pleasure to help clean up some of the little mess left after all the heavy lifting has been done.

((Since reading the ["Frescobaldi?" lilypond-user thread](https://lists.gnu.org/archive/html/lilypond-user/2024-04/msg00290.html) I was slowly accumulating courage and determination to attempt the PyQt upgrade, and it's hard to describe my joy when I found out that someone else has already done that - and that Frescobaldi is even freed from the dependency on PopplerQt, which was notoriously hard to maintain.))

Reply:
At first I only planned to use the upgrade branch for my daily LilyPond editing and fix the errors I run into, but then I tried running Frescobaldi with [coverage](https://coverage.readthedocs.io/), which resulted in an ongoing game of trying to reach as many hidden corners of the codebase as possible. I've always used only a very limited subset of Frescobaldi's functionality and it's interesting to see all the hidden features I never cared about.

Reply:
Yeah, discovering all those "hidden" features has been one of my favorite parts. Until I broke it the other week I had no idea "Cut and Assign" was a thing. That one's going to get a *lot* of use here.

Reply:
I didn't know "Cut and Assign" either. I might have discovered it earlier were it called "Extract variable", but that may be just a naming bias from the coding editors I happen to use the most.

Needless to say that both frescobaldi/qpageview#28 and frescobaldi/qpageview#29 are also features I had no idea about until last week.